src/main/kotlin/gg/essential/elementa/debug/InspectorManager.kt

```diff
@@ -0,0 +1,155 @@
+     */
+    fun toggleInspectorDetached(window: Window) {
+        val session = debugSessions[window] ?: return
+
+        session.isExternal = !session.isExternal
+
+        session.components.filterIsInstance<Inspector>().forEach { inspector ->
+            inspector.setFloating(false) // To prevent it from lingering as a floating component
+        }
+
+        if (session.isExternal) {
+            session.components.forEach {
+                window.removeChild(it)
+                session.display.addComponent(it)
+            }
+        } else {
+            session.components.forEach {
+                session.display.removeComponent(it)
+                window.addChild(it)
+            }
+        }
+    }
+
+    /**
+     * Called immediately following [Window.draw] with the current window
+     */
+    internal fun onDraw(window: Window) {
+        val openDisplay = debugSessions[window] ?: return
+        openDisplay.display.updateFrameBuffer(window.resolutionManager)
+    }
+
+    /**
+     * Cleans up the inspector and/or debug component from the supplied [window]
+     */
+    private fun cleanupWindow(window: Window) {
+        val inspectorState = debugSessions[window] ?: return
+        inspectorState.components.filterIsInstance<Inspector>().forEach { inspector ->
+            inspector.cleanup()
+        }
+        inspectorState.display.cleanup()
+        debugSessions.remove(window)
+    }
+
+    /**
+     * Adds a custom debug component either directly to [window] or to an external display
+     * depending on whether the current session is external or not. If no debug instance exists,
+     * then a new one will be created from on the current value of [startDetached]
+     */
+    fun addCustomDebugComponent(window: Window, component: UIComponent) {
+        val openDisplay = debugSessions[window]
+        if (openDisplay == null) {
+            createExternalDisplay(window, component)
+        } else {
+            openDisplay.components.add(component)
+
+            if (openDisplay.isExternal) {
+                openDisplay.display.addComponent(component)
+            } else {
+                window.addChild(component)
+            }
+        }
+    }
+
+    /**
+     * Removes a custom debug component from the debug session of the supplied [window].
+     * If the result of this operation leaves no debug components, then the debug session is closed.
+     *
+     * If no debug session exists, then nothing will be done.
+     */
+    fun removeCustomDebugComponent(window: Window, component: UIComponent) {
+        val session = debugSessions.remove(window) ?: return
+        session.components.remove(component)
+        if (session.isExternal) {
+            session.display.removeComponent(component)
+        } else {
+            window.removeChild(component)
+        }
+        if (session.components.isEmpty()) {
+            cleanupWindow(window)
+        }
+    }
```

Comment:
How's this supposed to work? If it's just added to the external display, then won't it necessarily get in the way of the inspector and vice versa? Would it not make more sense to give custom components their own window, or a dedicated space in the existing inspector UI? Do we even need to support custom debug components? That's one ginormous behavioral api surface we're exposing if we do, I would very much prefer we don't; at least we should put it behind a OptIn so we don't have to provide backwards compatibility.